### PR TITLE
[s]Fixes RPD/disposals exploits

### LINKED ISSUE #13473: RPD can be used to make infinite obstacles for no cost
A gang used this to stump the crew for 800 seconds.
sybil expensive station x-ray 2015-12-06 020550
pls nerf

### LINKED ISSUE #14654: RPDs can place pipes on top of unbreakable walls
this means you can abuse them with disposals rides to traverse through "unpassable" walls in away missions etc.

### PR DESCRIPTION
**Pull request by tkdrg**
Also removes a bit of copy-paste and speeds up a helper slightly.

Fixes #13473
Fixes #14654

### CODE REVIEW

File: code/game/objects/items/weapons/RPD.dm [hunk 1]
```diff
@@ -529,6 +529,9 @@ var/global/list/RPD_recipes=list(
 
 	if(istype(A,/area/shuttle)||istype(A,/turf/space/transit))
 		return 0
+	
+	if(is_blocked_turf(get_turf(A))) // This used to be allowed to help people rebuild pipes destroyed under walls etc
+		return 0 // But some people exploited it and forced me to do this
```
Comment (duncathan) [inline code/game/objects/items/weapons/RPD.dm]: Disgusting. At least only make it apply to disposals pipes.

Comment (ChangelingRain) [inline code/game/objects/items/weapons/RPD.dm]: Or at least only apply this to walls, is_blocked_turf() will fail on dense mobs, tables, random objects, and honestly a disgusting amount of stuff

Comment (tkdrg) [inline code/game/objects/items/weapons/RPD.dm]: Well duh, the pipe exploit could be used to get past any dense object. I could change it to check for both density and anchored, I suppose.

File: code/game/objects/items/weapons/RPD.dm [hunk 2]
```diff
@@ -529,6 +529,9 @@ var/global/list/RPD_recipes=list(
 
 	if(istype(A,/area/shuttle)||istype(A,/turf/space/transit))
 		return 0
+	
+	if(is_anchored_dense_turf(get_turf(A))) // This used to be allowed to help people rebuild pipes destroyed under walls etc
+		return 0 // But some people exploited it and forced me to do this
```
Comment (duncathan) [inline code/game/objects/items/weapons/RPD.dm]: Make it apply only to disposals pipes

File: code/modules/recycling/disposal-construction.dm [hunk 3]
```diff
@@ -224,9 +220,7 @@
 			anchored = 1
 			if(ispipe)
 				level = 1 // We don't want disposal bins to disappear under the floors
-				density = 0
-			else
-				density = 1 // We don't want disposal bins or outlets to go density 0
+			density = 0
```
Comment (phil235) [inline code/modules/recycling/disposal-construction.dm]: Are you sure this doesn't make disposal bin not dense like the comment above says?

Comment (tkdrg) [inline code/modules/recycling/disposal-construction.dm]: it does, that's intended

File: code/game/objects/items/weapons/RPD.dm (hunk 2 above)
Comment (KazeEspada) [inline code/game/objects/items/weapons/RPD.dm]: Don't walls have a var for indestructable?

Comment (Shadowlight213) [inline code/game/objects/items/weapons/RPD.dm]: duncathan I don't see that feature used to much when rebuilding, as the resulting pipes end up layared above the wall.

Comment (duncathan) [inline code/game/objects/items/weapons/RPD.dm]: I have yet to play a full non antag round in which I haven't used that feature

Comment (Shadowlight213) [inline code/game/objects/items/weapons/RPD.dm]: Must be just me then. The pipes layered on top of the walls trigger ne so I actively avoid using it.

Comment (duncathan) [inline code/game/objects/items/weapons/RPD.dm]: add this check to line 586 instead.
`if(!istype(A, /turf) || is_anchored_dense_turf(A))`

**Pull request merged**